src/conversion.rs

```diff
@@ -237,6 +237,61 @@ pub trait FromPyObject<'py>: Sized {
     }
 }
 
+/// Expected form of [`FromPyObject`] to be used in a future PyO3 release.
+///
+/// The difference between this and `FromPyObject` is that this trait takes an
+/// additional lifetime `'a`, which is the lifetime of the input `Bound`.
+///
+/// This allows implementations for `&'a str` and `&'a [u8]`, which could not
+/// be expressed by the existing `FromPyObject` trait once the GIL Refs API was
+/// removed.
+///
+/// # Usage
+///
+/// Users are generally advised against implementing this trait, instead they should
+/// implement the normal `FromPyObject` trait. This trait has a blanket implementation
+/// for `T: FromPyObject`.
+///
+/// The only case where this trait should be implemented is when the lifetime of the
+/// extracted value is tied to the lifetime `'a` of the input `Bound` instead of the
+/// GIL lifetime `py`, as is the case for the `&'a str` implementation.
+///
+/// Similarly, users should typically not call these trait methods and should instead
+/// use this via the `extract` method on `Bound` and `Py`.
+pub trait FromPyObjectBound<'a, 'py>: Sized {
+    /// Extracts `Self` from the bound smart pointer `obj`.
+    ///
+    /// Users are advised against calling this method directly: instead, use this via
+    /// [`Bound<'_, PyAny>::extract`] or [`Py::extract`].
+    fn from_py_object_bound(ob: &'a Bound<'py, PyAny>) -> PyResult<Self>;
```

Comment:
We could also already call this `from_py_object`, but maybe it's better to have a clear distinction for now...

Reply:
I think the one question which made me hesitate on encouraging users to use this trait too much (and also make this name clunky) is that I think to resolve the `TODO` in instance.rs:

```
        // TODO it might be possible to relax this bound in future, to allow
        // e.g. `.extract::<&str>(py)` where `py` is short-lived.
        'py: 'a,
```

we would need to make this trait accept `ob: Borrowed<'a, 'py, PyAny>` as the input, as that's basically `&Bound` with additional flexibility on the relationship between `'a` and `'py`.

It felt to me that this is a hurdle we can cross down the line, and working with `Bound` for now keeps it simple...

Reply:
Ah, using `Borrowed` there would certainly be an option. I agree that this is a question best answered once we actually removed the gil-refs and do the final migration. We could also seal this trait for now (but maybe that's a bit drastic). Removing a seal in a patch release, if it turns out to be necessary, would also be an option.

Reply:
Ok, I pushed an additional commit to seal the trait.